File: src/lib/api/trollstore.ts
import { invoke } from '@tauri-apps/api/core';

export interface TrollStorePrepareRequest {
  savedDir: string;
  forceVersion?: string | null;
}

export interface TrollStorePrepareResult {
  version: string;
  tarPath: string;
  helperPath: string;
  cached: boolean;
}

export interface TrollStoreEligibilityRequest {
  productType: string | null;
  iosVersion: string | null;
}

export type TrollStorePath =
  | 'ios14-15-ramdisk'
  | 'ios16-trollrestore'
  | 'incompatible'
  | 'unknown';

export interface TrollStoreEligibilityResult {
  path: TrollStorePath;
  reason: string;
  iosMajor: number | null;
}

export function prepareTrollstoreAssets(
  request: TrollStorePrepareRequest,
): Promise<TrollStorePrepareResult> {
  return invoke<TrollStorePrepareResult>('prepare_trollstore_assets', { request });
}

export function checkTrollstoreEligibility(
  request: TrollStoreEligibilityRequest,
): Promise<TrollStoreEligibilityResult> {
  return invoke<TrollStoreEligibilityResult>('check_trollstore_eligibility', { request });
}

File: src-tauri/src/models/trollstore.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrollStorePrepareRequest {
    pub saved_dir: String,
    /// Optional override for testing — if Some, skip GitHub API call and use this tag directly.
    pub force_version: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TrollStorePrepareResult {
    pub version: String,
    pub tar_path: String,
    pub helper_path: String,
    pub cached: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrollStoreEligibilityRequest {
    pub product_type: Option<String>,
    pub ios_version: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
pub enum TrollStorePath {
    #[serde(rename = "ios14-15-ramdisk")]
    Ios14To15Ramdisk,
    #[serde(rename = "ios16-trollrestore")]
    Ios16PlusTrollRestore,
    #[serde(rename = "incompatible")]
    Incompatible,
    #[serde(rename = "unknown")]
    Unknown,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TrollStoreEligibilityResult {
    pub path: TrollStorePath,
    pub reason: String,
    pub ios_major: Option<u32>,
}
